inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -774,15 +818,225 @@ static void chanToPlaneRow(const uint8_t* in, int chan, int chs, uint8_t* out, i
+    using p_f = void (*)(const uint8_t** y_rows, const uint8_t* uv_row,
+        uint8_t** out_rows, const int buf_width);
+
+    template <typename type>
+    p_f operator()(type_to_type<type>) {
+        return [](const uint8_t** y_rows, const uint8_t* uv_row,
+            uint8_t** out_rows, const int buf_width) {
+                const auto inT1 = reinterpret_cast<const type**>(y_rows);
+                const auto inT2 = reinterpret_cast<const type*>(uv_row);
+                auto outT = reinterpret_cast<type**>(out_rows);
+
+                nv12ToRgbRowImpl(isa_tag_t{}, inT1, inT2, outT, buf_width);
+        };
+    }
+};
+}  // namespace
+
+namespace {
+
+using i420_to_rgb_supported_types = typelist<uint8_t>;
+
+static void i420ToRgbRowImpl(scalar_tag, const  uint8_t** y_rows,
+    const  uint8_t* u_row,
+    const  uint8_t* v_row,
+    uint8_t** out_rows,
+    const int buf_width) {
+    for (int i = 0; i < buf_width; i += 2) {
+        uchar u = u_row[i / 2];
+        uchar v = v_row[i / 2];
+        int ruv, guv, buv;
+        uvToRGBuv(u, v, ruv, guv, buv);
+
+        for (int y = 0; y < 2; y++) {
+            for (int x = 0; x < 2; x++) {
+                uchar vy = y_rows[y][i + x];
+                uchar r, g, b;
+                yRGBuvToRGB(vy, ruv, guv, buv, r, g, b);
+
+                out_rows[y][3 * (i + x)] = r;
+                out_rows[y][3 * (i + x) + 1] = g;
+                out_rows[y][3 * (i + x) + 2] = b;
+            }
+        }
+    }
+}
+
+template<typename isa_tag_t>
+struct typed_i420_to_rgb_row {
+    using p_f = void (*)(const uint8_t** y_rows, const uint8_t* u_row, const uint8_t* v_row,
+                         uint8_t** out_rows, const int buf_width);
+
+    template <typename type>
+    p_f operator()(type_to_type<type>) {
+        return [](const uint8_t** y_rows, const uint8_t* u_row, const uint8_t* v_row,
+                  uint8_t** out_rows, const int buf_width) {
+                const auto inT1 = reinterpret_cast<const type**>(y_rows);
+                const auto inT2 = reinterpret_cast<const type*>(u_row);
+                const auto inT3 = reinterpret_cast<const type*>(v_row);
+                auto outT = reinterpret_cast<type**>(out_rows);
+
+                i420ToRgbRowImpl(isa_tag_t{}, inT1, inT2, inT3, outT, buf_width);
+        };
+    }
+};
+}  // namespace
+
+template <typename isa_tag_t>
+struct choose_impl {
 GAPI_FLUID_KERNEL(FChanToPlane, ChanToPlane, false) {
     static const int Window = 1;
     static void run(const cv::gapi::fluid::View& in, int chan,
                     cv::gapi::fluid::Buffer& out) {
-        const auto rowFunc = (in.meta().depth == CV_8U) ? &chanToPlaneRow<uint8_t> : &chanToPlaneRow<float>;
+        GAPI_DbgAssert(is_cv_type_in_list<chan_to_plane_supported_types>(out.meta().depth));
+
+        const auto rowFunc = type_dispatch<chan_to_plane_supported_types>(out.meta().depth, cv_type_id{}, typed_chan_to_plane_row<isa_tag_t>{}, nullptr);
```

Comment:
so do we still dispatch in EVERY row? Or is the ISA-based dispatch done statically here?

Reply:
Yes, we do ISA-based dispatch in every row. This solution was proposed by Anton in the ChanToPlan kernel. Anton said that he investigated possibility to do that via compound kernel and he found out that we can't do that.